### ISSUE TITLE
ESP8266 LED being overridden by FastLED

### ISSUE THREAD

**Issue by LeafyDev** [FastLED/FastLED#549]
Hi, sorry if this is addressed somewhere else, but I have a NodeMCU using WiFi and FastLED to control some Neopixel strips, I make the onboard LED flash (pin 2) during setup phase, but then when I tell it to turn off, it won't ever stay off...

If I remove FastLED from the includes (and therefore all the code that uses it) the LED behaves fine. Yet I've looked through the FastLED code and see that the only place it could be used is the gMaxPowerIndicatorLEDPin... Which I've played with with no success...?

Any ideas? Thanks

**Comment by focalintent** [comment]
@kriegsman - know what's up with this?

**Comment by kriegsman** [comment]
Huh.  I think I have a NodeMCU around here, I'll see if I can check.  
@LeafyDev: Exactly which NodeMCU board are you using?  I think mine is the older NodeMCU 0.9 revision.  I'd be happy to get a new board and test, but I'd like to know precisely which NodeMCU board you have there.  Thanks!

**Comment by FaldoAU** [comment]
Just poking my head in to say: you guys know you're the fucking best,
right?!

**Comment by LeafyDev** [comment]
It's a Chinese one I got from Amazon, it exactly looks like a 1.0 and is on Amazon (France) with this name:

Bubudong ESP8266 Serial Port Module Wifi NodeMcu Lua WIFI V3

I'm available to test stuff and screen share with webcam to avoid you buying stuff if needed

**Comment by kriegsman** [comment]
Thanks for that info -- I wanted to make sure that I had a board that matched the hardware you have exactly.  Luckily, they're so cheap and so easy to get that I should have my board to test with in just a couple of days.

For reference, this was the image on amazon.fr showing the board... note the rectangular USB/serial converter (CH340) and the LoLin silkscreening.
https://images-na.ssl-images-amazon.com/images/I/61pXFl9MTuL._SL1020_.jpg
and here's the one that I have heading to me for testing
https://images-na.ssl-images-amazon.com/images/I/71XlmUW%2BQDL._SL1200_.jpg
Looks very similar.

In contrast, here are some (newer?) ones with a square USB/serial converter and no LoLin silkscreening:
https://images-na.ssl-images-amazon.com/images/I/71GUeTwFn5L._SL1088_.jpg

So we'll see what actually arrives, but I appreciate this ticket report.  I'll let you know what I find once I have hardware and some time to test it.  Thank you!

**Comment by LeafyDev** [comment]
Oh wow, thanks!

I'm adding an image to show you why this bothers me, it's insanely bright :(

![Bright LED](https://cdn.discordapp.com/attachments/239588543167463424/402811353368100864/Bp6iAE9xs4_oPA6ABVExGcFT_T4JDY30-Lsz8JRRObSvUrhFI1wDBQTXLMIuhTv0QEfLhCKLSLILo03mJ3qMcTjF6hy2aH7Vvo_1.png)

And also the code I use: https://gist.github.com/LeafyDev/e9ff8a4c8ec09ec1604c4fb764626b1a

Let me know if you need anything else :)

**Comment by kriegsman** [comment]
Yeah... that's brighter than one might wish.  Thanks for the code and I'll let you know what I find.

**Comment by marmilicious** [comment]
Commenting because I'm curious to know what you hopefully discover on this strange one Mark.  

@LeafyDev Yes, those little onboard LEDs are sometimes super bright!  I did use that to my advantage once though, sending it out a little custom light pipe to the outside of a case.
https://www.youtube.com/watch?v=TKM7vcbbOy0&feature=youtu.be&t=21s

**Comment by pacobyte** [comment]
@LeafyDev Here's something you can try in the meantime. I use FastLED with a NodeMCU, but mine is what I think is referred to as v2 (it looks more like the last linked image in kriegsman's post above). My board has **two** built-in LEDs—a red one on the dev board and a blue one on the actual ESP8266 (which appears to be the same as the one that's lit in your posted photo).

In order to use the red LED, I simply need to refer to it as `LED_BUILTIN` with no need to `#define` it. So, to turn it on would just be:
``` c
pinMode(LED_BUILTIN, OUTPUT);
digitalWrite(LED_BUILTIN, LOW);
```
I've heard that `LED_BUILTIN` is an auxiliary constant and not a board pin, so if you also have that 2nd LED on your NodeMCU, maybe that one would play nicer with FastLED.

A workaround, I know, but I thought I'd throw that out there.

**Comment by LeafyDev** [comment]
I have looked into this yes, but either the LED isn't on my board or
doesn't respond to LED_BUILTIN stuff, I may give it another go this evening
without including any libraries just to see... Thanks for the idea though.

**Comment by LeafyDev** [comment]
Hi, quick update, I might be missing something but it may not actually be FastLEDs fault, since this project uses it and I don't get any LED activity at all...

https://github.com/jasoncoon/esp8266-fastled-webserver/

It may be a different library messing with it, but if I comment out FastLEDs includes, it works perfectly as expected.. Interesting

**Comment by connornishijima** [comment]
You wouldn't by chance be running FastLED on the GPIO that the built-in LED is tied to, right?

**Comment by chrisparton1991** [comment]
You may have solved this by now, but instead of reading the state with `int state = digitalRead(LED_PIN)`, can you try using a global boolean to store the LED state and use that?

Given that the pin is set to output and it has an LED attached, it's possible that reading the pin is always returning LOW.

**Comment by m2049r** [comment]
`FastLED.show()` just turns the LED on. Interestingly, if the LED is disabled with `pinMode(LED_BUILTIN, INPUT)` then `FastLED.show()` does nothing!

**Comment by zackees** [comment]
Closing as **not reproducible** after seven years — but with the most likely explanations written down, since this thread never got a conclusion.

@kriegsman ordered a matching LoLin board in Jan 2018 to chase it and the thread went quiet after that; there was never a confirmed reproduction or a fix. @MoonieGZ himself later noted "it may not actually be FastLED's fault."

Three things in the thread are probably the whole story, and any of them would produce exactly what you saw:

**1. `digitalRead()` on an OUTPUT pin.** @chrisparton1991's last comment is the strongest lead and was never tested. Reading back a pin configured as OUTPUT with an LED on it does not reliably return what you wrote, so a toggle written as `digitalWrite(pin, !digitalRead(pin))` can get stuck asserting one state forever. Keep the state in a `bool` and write that instead. This is a pure sketch-side bug and would explain why the *blinking* worked (unconditional writes) while the *turn-it-off* step didn't (read-modify-write).

**2. GPIO2 is heavily contended on the ESP8266.** @connornishijima asked whether FastLED was driving the same GPIO — worth restating, because on this chip pin 2 is simultaneously the on-module blue LED, a boot-strapping pin that must be HIGH at reset, and UART1 TX. Anything that touches UART1 drives that LED as a side effect. That is also why the LED is *inverted* on most of these modules: LOW turns it on.

**3. Two different LEDs.** @pacobyte's point stands — the red LED on the dev board and the blue LED on the ESP8266 module itself are separate, on separate pins, and `LED_BUILTIN` refers to only one of them depending on the board variant.

**Relevant to anyone landing here today:** FastLED now ships a UART-based WS2812 driver for the ESP8266 that transmits on **UART1 / GPIO2** (`src/platforms/esp/8266/fastled_esp8266_uart.h`). If that's enabled, GPIO2 belongs to the LED driver and the on-board LED will do whatever the pixel data says. It is strictly opt-in — it requires `#define FASTLED_ESP8266_UART` before including `<FastLED.h>`, or adding `UARTController_ESP8266` explicitly — so it cannot be the cause of a 2018 report, but it is a real GPIO2 conflict to be aware of now.

If anyone still reproduces this on a current FastLED with a sketch that keeps LED state in a variable and does not use GPIO2 for pixel data, please open a fresh issue with that sketch — it'd be a genuinely new finding rather than a continuation of this one.

Thanks @MoonieGZ for the detailed report and the offer to screen-share for debugging, and @pacobyte, @chrisparton1991, @connornishijima, @marmilicious for the diagnostics.
